doc/i2c_bus_lock_prevention/i2c_bus_lock_prevention_hld.md

```diff
@@ -0,0 +1,425 @@
+
+###### Table 2: Abbreviation
+| Abbreviation              | Description                                            |
+|---------------------------|--------------------------------------------------------|
+| I2C                       | Inter integrated circuit                               |
+
+### 4. Overview 
+
+Many peripheral devices relies on the i2c bus to operate. For example, transceivers, thermal sensors and psu modules.
+When the i2c bus is locked, no transactions can be done through the i2c. The temperatures of transceivers, the temperatures
+from thermal sensors and psu modules are not able to be retrieved. It is possible that the control of the fan speed
+also relies on i2c and it is not able to control the fan speed when the i2c bus is locked. That means, the thermal policy of
+the system cannot be applied and the hardware device could be in the risk of damage due to over temperature.
+
+This feature introduces a mechanism to monitor the i2c bus health status.
+When the i2c bus is locked, the daemon detect the faulty device on the i2c bus and generate an isolation list which
+contains the detected faulty devices. All of the daemons which access i2c bus will be stopped during the
+aforementeiond operation. After the isolation list is generated and the faulty device is marked through the platform
+API, the daemons which access i2c bus will be started. After the daemons are started, the function calls by the daemons
+to the platform API will avoid initating an i2c transaction to the faulty device so that the i2c bus will not be locked
+again.
+
+The hardware related requirements, the concept of a i2c region, the mechanism to detect i2c bus lock and recover
+the i2c bus lock, the platform driver and platform API related changes, and the syslog messages written by this
+feature will be expalined in the following sections.
+
+### 5. Requirements
+
+Feature Requirements
+* Detect the i2c bus lock condition automatically.
+* When i2c bus lock happens, detect the faulty i2c devices and put the faulty devices into an isolation list.
+* Recover the i2c bus from lock status by preventing any program to access the faulty devices in the isolation list.
+* The design shall not impact system performance, or minimize the impact to the system performance.
+* Data plane traffic shall not be affected after the i2c recovery process is applied unless the conditions explicitly mention below.
+    - If the faulty device is a transceiver inserted on a port which has running traffic, the traffic will be affected when it is replaced with a good device.
+
+Hardware Requirements
+* This feature requires the hardware support to close all channels of a I2C mux through non-i2c interface(e.g. CPU's GPIO).
+* The swappable device shall be located behind a channel of a I2C mux.
+
+
+### 6. Architecture Design 
+
+#### 6.1. I2C Healthd

```

Comment:
Is there an advantage to having a separate daemon polling i2c devices? I'm concerned about contention, especially on platforms that have BMC (baseboard management controllers) that may also perform or manage i2c access. 

An alternative approach is to add the i2c bus lock detection capability into the individual platform daemons that already have exclusive or almost-exclusive control over the i2c devices. Then we wouldn't run the risk of competing for bus access. For example, `xcvrd` for transceivers, `thermalctld` for temp sensors and fans, `psud` for PSUs, etc. These daemons can manage and sequence their access to i2c, perhaps in a multi-threaded approach. Was this approach considered? If so, please add a section explaining why it wasn't selected.